Add Db queries verification to  tfb toolset verif procedure

jcheron opened the pull request:
It seems necessary to know if each tested framework or platform always responds in case of stress (high concurrency level) and actually perform all queries to the database.
The [Ubiquity-workerman case ](https://github.com/TechEmpower/FrameworkBenchmarks/issues/5118) proved that a framework can silently fail (no logs, no errors issued when not all db queries are made).

I guess the Techempower team has all the skills to make this change, so if I could just save you some time.

This PR proposes a solution to evaluate the number of queries actually made, at the maximum concurrency level (512), and integrates this counting into the `verification` phase:

```
./tfb --mode verify --test frameworkName 
```
For each test requesting the database (**Db, Queries, Updates, Fortunes**), the principle is as follows:

0. Checking of failed transactions (0 expected) -> except on Travis-ci
1. Retrieval from the database statistics of the number of queries made, the number of rows read (and eventually updated).
2. Realization of `n` http requests at the concurrency level 512 with a bench tool (`siege`).
3. Retrieving statistics again, calculating the number of queries made, the number of rows read.
4. Comparison with expected values, and failure in case of deviation (insufficient number).
A margin of 1% is acceptable for the lines read (it seems to me that the stats are not 100% reliable)

**Note:**
The acceptance margin on db queries is higher for Travis-ci runs, as resources are limited.

The method can certainly be improved, this is a first proposal...

Having tested quite a few frameworks for development purposes, it seems that Ubiquity workerman is not the only one in his case.

jcheron commented:
### Example of results with the new verification procedure
#### Problematic Ubiquity-workerman version
![image](https://user-images.githubusercontent.com/2511052/66648127-11ecef00-ec2b-11e9-9ee2-e365fbf14dde.png)

![image](https://user-images.githubusercontent.com/2511052/66647908-752a5180-ec2a-11e9-9d8a-e17b110acdfe.png)

### Updated Ubiquity-workerman version
![image](https://user-images.githubusercontent.com/2511052/66648407-bc651200-ec2b-11e9-802f-eff72364da17.png)

nbrady-techempower commented on toolset/benchmark/test_types/framework_test_type.py:
```diff
@@ -129,3 +134,89 @@ def copy(self):
         '''
         return copy.copy(self)
+
+    def get_current_world_table(self):
```
This is removed in #4549. We don't want any database specific stuff in the toolset. They should be contained to the `databases` folder.

nbrady-techempower commented on toolset/benchmark/test_types/verifications.py:
```diff
@@ -384,4 +393,42 @@ def verify_query_cases(self, cases, url, check_updates=False):
+    problems = []
+    dbChecker=None
+
+    if dbType == "mysql":
```
This as well. You can make a `verify_queries` method on the `Database` class inside each database folder and then call it here instead of checking for the proper db type. Something like:

```
from toolset.databases import databases

databases[dbType].verify_queries(self.config)
```

jcheron replied:
I must necessarily add several methods in each `Database` class, only one `verify_queries` method would not be very clean (see https://github.com/TechEmpower/FrameworkBenchmarks/pull/5145/commits/87358fb0bc5c10d69c0d49c7e377f6c0341e35c5)

msmith-techempower replied:
I think he's saying make a new abstract method on the parent class called `verify_queries` and then each subtype would implement it their own way, so instead of
```Python
if dbType == "mysql":
  dbChecker=MysqlCheck(self.config)
elif dbType == "postgres":
  dbChecker=PostgresCheck(self.config, tbl_name)
elif dbType == "mongodb":
  dbChecker=MongoDbCheck(self.config)
```
you could just have what he wrote:
```Python
databases[dbType].verify_queries(self.config)
```
as `databases[dbType]` would be the subclass and therefore `verify_queries` would be the type-specific implementation - if `databases[dbType]` was a MySQL type, then it's implementation of `verify_queries` would be whatever `MysqlCheck(self.config)` does.

jcheron replied:
ok, but `verify_queries` is common to all Database classes:
Overall, she does that:
- getQueries & getRows
- do `ab`
- getQueries & getRows
- compare queries and row count with expected values

These are the `getQueries` and `getRows` methods that are specific to a subclass

jcheron replied:
`verify_queries` added in a Database super class: see https://github.com/TechEmpower/FrameworkBenchmarks/pull/5145/commits/a5c826a028c8cef31e8465af8cc695d1aeef494c

jcheron commented:
> Unfortunately I think you just missed the merge of #4549, but it shouldn't be too hard to make this fit with that.

Yes, it's unfortunate!
I therefore merge my Db classes with those of this [PR](https://github.com/TechEmpower/FrameworkBenchmarks/pull/4549)?

nbrady-techempower commented:
Yes, that should pretty much do it and then we can clean up from there.

sebastienros commented on toolset/benchmark/test_types/framework_test_type.py:
```diff
@@ -128,4 +128,4 @@ def copy(self):
         Use before calling parse
         '''
-        return copy.copy(self)
+        return copy.copy(self)
```
nit

```suggestion
        return copy.copy(self)

```

jcheron commented:
Sounds good to me, and ready for clean up.

`ApacheBench` has been replaced by `Siege` (Http 1.1 compatible) for the sending of requests (2 x 512 for each url tested).

`Mysql` is quite imprecise for statistics on read or updated rows: it tends to underestimate them.
So I had to introduce a margin of error for acceptance.
No problems with other databases.

I did not find any inconsistencies in the results that I looked at closely.

joanhey commented:
After see the check, some frameworks use a lot of more queries or/and rows read, and are marked as `passed`.
I think will be labeled as `warm`, so they could check that anomaly.

jcheron commented:
> After see the check, some frameworks use a lot of more queries or/and rows read, and are marked as `passed`.
> I think will be labeled as `warm`, so they could check that anomaly.

I was able to add more precision to the calculation of Mysql queries.
There should no longer be such a large gap now, so I am also in favour of adding a warning if the results exceed a certain percentage.

msmith-techempower commented on toolset/databases/abstract_database.py:
```diff
@@ -0,0 +1,95 @@
+        cls.reset_cache(config)
+        #Start siege requests
+        path = config.db_root
+        os.system("siege -c %s -r %s %s -R %s/.siegerc" % (concurrency, count, url, path))
```
How is this better than making a single HTTP request to the server?

Is the purpose of this to identify implementations which are making good-faith attempts at queries, but failing [somehow] and therefore getting better, albeit fake, performance?

jcheron replied:
> How is this better than making a single HTTP request to the server?

It's a bit of an arbitrary choice,
maybe in case the second repetition reveals something other than the first?
`repetition` is an adjustable variable.

> Is the purpose of this to identify implementations which are making good-faith attempts at queries, but failing [somehow] and therefore getting better, albeit fake, performance?

Yes, what matters is to identify the non-compliance with the test specifications, there is an obligation of means, but also an obligation of result, whatever the performance (better or not).


For the moment, I have identified three possible cases causing failures:
- non-compliance with the code (for example, updating without prior reading)
- failed Http requests
- connections to the database probably impossible due to overload (sometimes causing errors, but not always)

We can possibly add stats on the number of `connections` to the database, to see if it provides answers.

nbrady-techempower commented:
@jcheron can you make another arbitrary commit and remove `[ci skip]` i'd like to see this make a full run

jcheron commented:
> @jcheron can you make another arbitrary commit and remove `[ci skip]` i'd like to see this make a full run

It seems the Docker network is down.
waiting for it to come back....

jcheron commented:
`Swoole-no-async` took over from `Ubiquity-workerman`:
![image](https://user-images.githubusercontent.com/2511052/66876637-d77cac80-efb2-11e9-9c3f-54c7a601d45a.png)
see [travis result](https://travis-ci.org/TechEmpower/FrameworkBenchmarks/jobs/598289484#L3729)
![image](https://user-images.githubusercontent.com/2511052/66876783-54a82180-efb3-11e9-82d5-1b2c9f015f49.png)

jcheron commented:
I'm trying to review the failed verifications.
And I think there must be an imperfection in the calculation of the number of queries with **PostgreSQL**.
For a number of frameworks with pg, on the `Update` test,
the verification finds `10752` db queries instead of `20480` normally (select+update)=>There are `9728` queries missing each time.
If I double the Http requests, the results remain in the same proportions (`21504` instead of `40960`)
The strange thing is that not all frameworks with postgreSQL do.
see [postgre.py](https://github.com/phpMv/FrameworkBenchmarks/blob/toolset-queries-verification/toolset/databases/postgres/postgres.py#L59)

jcheron commented:
@nbrady-techempower 
I think we're not bad here.
The Travis results clearly show frameworks that have problems at the requested concurrency level (socket errors, overloaded database connection...), and that could have gone before without being clearly identified: cases like **Ubiquity-workerman**, or **Swoole-no-async** more recently.

Some informations about the stats:
#### queries
The number of executed queries returned by the statistics is accurate and seems reliable.
I didn't put a margin on it.
I have taken into account possible bulk queries on the update. If they are used, the number of expected requests is reduced accordingly.
#### rows
- For the numbers of rows read or updated, the statistics are a little less reliable. I have set a default acceptance margin of 1% (1.5% for Mysql).
- In case of Bulk queries, I added an additional margin (5%) certainly corresponding to the random aspect of the update.
- I have added a warning if the expected values are significantly exceeded.

All these margins are of course adjustable.

Last but not least, Travis' results are certainly distorted, compared to what we can get on `Citrine` (the resources are not the same).

nbrady-techempower commented:
Hey @jcheron,

Thank you for your continued work on this!

I'm looking at things like this in https://travis-ci.org/TechEmpower/FrameworkBenchmarks/jobs/600132782#L852 in the latest run:

```
   FAIL for http://tfb-server:8080/update?queries=20
     Only 20468 executed queries in the database out of roughly 20480 expected.
     See https://github.com/TechEmpower/FrameworkBenchmarks/wiki/Project-Information-Framework-Tests-Overview#specific-test-requirements
```

I wouldn't expect this to be a failure if we're allowing for a 1% margin of error? (1.5% on mysql)

jcheron commented:
Hi @nbrady-techempower ,
I didn't notice that the calculation of the number of queries was not reliable  (unlike the statistics on rows), so I didn't apply any margin to it.
The margin applies only to rows read or updated.
For the case mentioned, I think that the 12 queries are missing (this is confirmed by the number of rows read), but this value is marginal !

We could eventually apply the same margin on queries, and to systematically display a warning in case of a lower value than expected, provided that it is within the acceptable margin.

jcheron commented:
@nbrady-techempower 
It's ok with a random selection of **unique** numbers in **ninja-standalone**:
![image](https://user-images.githubusercontent.com/2511052/67576022-0bd43380-f73e-11e9-91c9-80b6dcd82494.png)

I believe it is more appropriate for frameworks using `hibernate` to do this modification, for reasons of equity, but especially because if we add a margin on the number of Db queries, we may no longer correctly identify the frameworks that really have problems.
The verification should allow the developer to know if what he submits is in conformity with the specifications.

nbrady-techempower commented:
@jcheron Yeah, I agree with you. We're discussing internally how we want to approach the Hibernate ones. This is looking really good. I don't think there's much else to do here but I want to go through and look at each new failure compared to the last master travis run.

jcheron commented:
Yes, this review of the last Travis run is necessary.
But it should not be forgotten that the Travis run is not representative of Citrine/Azure benchmarks.

This verification procedure requires resources (memory, processor)  a little light on Travis and some frameworks (especially those that work asynchronously) do not pass Travis tests, whereas they will certainly pass benchmarks (this is the case for Laravel-Swoole, Swoole...).

nbrady-techempower commented:
> But it should not be forgotten that the Travis run is not representative of Citrine/Azure benchmarks.

Yes, I know this, however, we need to make sure that Travis runs represent the minimum success for a framework. If they don't pass Travis they're not getting merged, so, we may have to make some adjustments. What is it about  Laravel-Swoole that it's missing 2 queries? Do we need to lower the concurrency from siege?

jcheron commented:
> What is it about Laravel-Swoole that it's missing 2 queries? Do we need to lower the concurrency from siege?

On Travis why not, but it would be unfortunate to do it on the real tests

It would be necessary to know what the verification is performed on (Travis or Citrine), and to be more tolerant on Travis (to which we add the margin),
Either we add a parameter at the input of the check (specifying if margin or not)
Either we recover the core number of the processor (if it is less than a certain value, we apply a margin)

nbrady-techempower commented:
@jcheron can you remove the `[ci skip]` from here on

nbrady-techempower commented:
@jcheron Could you merge in the latest master and give this a full run. I think we're about ready to get this in so that new PR's will test against it while we're fixing other frameworks.

jcheron commented:
@nbrady-techempower 
It may be more convenient if you have control over the initial repo.
https://github.com/phpMv/FrameworkBenchmarks/invitations

nbrady-techempower commented:
We're just gonna go for it.

nbrady-techempower commented:
I got this in because there was a new run on Citrine about to start, so let's see how this does there.

msmith-techempower commented:
> I got this in because there was a new run on Citrine about to start, so let's see how this does there.

I confirmed that it is included in the latest run on Citrine.

nbrady-techempower commented:
Good work team

jcheron commented:
> Good work team
yes to self-congratulation!
We have been efficient.

And for now, everything looks normal on Citrine

jcheron commented:
That's a little worrying....

![image](https://user-images.githubusercontent.com/2511052/68065053-d32af000-fd23-11e9-9c9f-0efcaae37449.png)

michaelhixson commented:
It looks like it crashed while verifying the updates test for akka-http-slick-postgres.  That framework's updates/verification.txt file is the most recently updated file that I see, and it has the "VERIFYING UPDATE" line but no results (such as "PASS") below that.  I don't see any stack traces or error messages in the logs.  The TFB application is not stuck; it exited.

I'm restarting it now, though I'm expecting it to get stuck in a similar way later tonight or tomorrow.

jcheron commented:
> It looks like it crashed while verifying the updates test for akka-http-slick-postgres. That framework's updates/verification.txt file is the most recently updated file that I see, and it has the "VERIFYING UPDATE" line but no results (such as "PASS") below that. I don't see any stack traces or error messages in the logs. The TFB application is not stuck; it exited.
> 
> I'm restarting it now, though I'm expecting it to get stuck in a similar way later tonight or tomorrow.

I will add a timeout on the `siege` command execution.

zloster commented on frameworks/Java/ninja-standalone/src/main/java/controllers/HelloDbController.java:
```diff
@@ -39,9 +40,12 @@ public Result multiGet(@Param("queries") Integer queries) {
-            worlds[i] = getRandomWorld();
-        }
+        //Pick unique random numbers 
+        final AtomicInteger i = new AtomicInteger(0);
```
ping @jcheron and @nbrady-techempower

michaelhixson replied:
@zloster If I were writing the code I'd also use `mapToObj`, but I'll share some findings about the performance that I thought were interesting.

Last week I tried using JMH to measure the overhead of different techniques for generating distinct world ids.  I just uploaded it to GitHub:  [repository](https://github.com/michaelhixson/jmh-benchmark-tfb-world-ids), [Java code](https://github.com/michaelhixson/jmh-benchmark-tfb-world-ids/blob/f4775e92f7715c2d8bebc605e445dd4e0ae5f0aa/src/main/java/rnd/WorldIdsBenchmark.java).

The results from my workstation are like this (lower is better):
```
Benchmark                        Mode  Cnt     Score    Error  Units
WorldIdsBenchmark.arrayCounter   avgt    5  1123.397 ± 44.659  ns/op
WorldIdsBenchmark.atomicCounter  avgt    5  1115.252 ± 50.794  ns/op
WorldIdsBenchmark.boxedSet       avgt    5   525.349 ± 15.149  ns/op
WorldIdsBenchmark.primitiveSet   avgt    5   375.261 ±  4.094  ns/op
WorldIdsBenchmark.streamOnly     avgt    5  1336.233 ± 27.164  ns/op
```

It seems that the overhead of additional stream ops in my preferred approach (`streamOnly`) is higher than the overheard from using `AtomicInteger` (`atomicCounter`).  But they are very close.  I feel like the impact of choosing any one approach over the other isn't going to make a noticeable difference in the context of TFB.  This piece of it affects around +/-1 microsecond per request (using the measurements from my workstation) and I imagine that's imperceptible next to all the database queries.

zloster replied:
@michaelhixson This is interesting, thank you. I see the JMH [test was run with `Java 13`](https://github.com/michaelhixson/jmh-benchmark-tfb-world-ids/blob/f4775e92f7715c2d8bebc605e445dd4e0ae5f0aa/pom.xml#L43). I'll check if the results are different with older JVM versions.
Note to myself: I don't like the too big difference between the basic loops and the streams variants.

jcheron replied:
My concern that day was not optimization... but when it is, I will tend to say that there is never any unnecessary optimization, even for a micro-second, even if I agree with @michaelhixson that it has no effect in the context of our db tests.
The benchmark is interesting, I wouldn't have bet on this result:
the simplest methods are often the best (+1 for `IntOpenHashSet`)

zloster replied:
@jcheron I use this as an [illustration how many CPU cycles are wasted these days](https://gist.github.com/zloster/bcf82dd558fc5733851455180fa37a49). Echoing 4 chars in a buffer is several instructions. The others (up to a ~0,7 million) instructions are very hefty price to achieve the goal. `JMH` has support for reporting similar data (instructions count). I'm curious to see the data.

zloster commented:
There is one more active issue [with similar problem](https://github.com/TechEmpower/FrameworkBenchmarks/issues/4391). I have suggested something similar [in a comment there](https://github.com/TechEmpower/FrameworkBenchmarks/issues/4391#issuecomment-459079062).
